File: ios/PoseLandmarkPayload.swift
import MediaPipeTasksVision

internal enum PoseLandmarkPayload {
  private enum Failure: Error { case mismatchedPoseCounts }

  static func make(_ result: PoseLandmarkerResult) throws -> [String: Any] {
    guard result.landmarks.count == result.worldLandmarks.count else {
      throw Failure.mismatchedPoseCounts
    }
    let poses = result.landmarks.enumerated().map {
      index, image -> (landmarks: [[String: Any]], worldLandmarks: [[String: Any]]) in
      let landmarks = image.map { joint -> [String: Any] in
        var coordinates: [String: Any] = ["x": joint.x, "y": joint.y, "z": joint.z]
        if let visibility = joint.visibility { coordinates["visibility"] = visibility }
        if let presence = joint.presence { coordinates["presence"] = presence }
        return coordinates
      }
      let worldLandmarks = result.worldLandmarks[index].map { joint -> [String: Any] in
        var coordinates: [String: Any] = ["x": joint.x, "y": joint.y, "z": joint.z]
        if let visibility = joint.visibility { coordinates["visibility"] = visibility }
        if let presence = joint.presence { coordinates["presence"] = presence }
        return coordinates
      }
      return (landmarks, worldLandmarks)
    }
    guard let first = poses.first else {
      return ["landmarks": [], "worldLandmarks": [], "poses": []]
    }
    return [
      "landmarks": first.landmarks, "worldLandmarks": first.worldLandmarks,
      "poses": poses.map { ["landmarks": $0.landmarks, "worldLandmarks": $0.worldLandmarks] },
    ]
  }
}

File: android/src/main/java/expo/modules/mediapipepose/PoseLandmarkPayload.kt
package expo.modules.mediapipepose

import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult

internal object PoseLandmarkPayload {
    fun make(result: PoseLandmarkerResult): Map<String, Any> {
        val landmarks =
            result.landmarks().firstOrNull().orEmpty().map { joint ->
                mutableMapOf<String, Any>("x" to joint.x(), "y" to joint.y(), "z" to joint.z())
                    .apply {
                        joint.visibility().ifPresent { put("visibility", it) }
                        joint.presence().ifPresent { put("presence", it) }
                    }
            }
        val worldLandmarks =
            result.worldLandmarks().firstOrNull().orEmpty().map { joint ->
                mutableMapOf<String, Any>("x" to joint.x(), "y" to joint.y(), "z" to joint.z())
                    .apply {
                        joint.visibility().ifPresent { put("visibility", it) }
                        joint.presence().ifPresent { put("presence", it) }
                    }
            }
        return mapOf("landmarks" to landmarks, "worldLandmarks" to worldLandmarks)
    }
}
